[GridBundle] [WIP] PHPCR-ODM Driver

dantleech opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | no
| BC breaks?      | no|yes
| Deprecations?   | no|yes
| Related tickets | 
| License         | MIT

This PR would introduce a PHPCR-ODM driver.

The main problem is building the query builder - the PHPCR-ODM query builder uses a [different design](https://gist.github.com/dantleech/298c9618b5889bddeb87) than that of the standard Doctrine Query Builder, which makes building the query more difficult.

In order to solve this this PR uses the Doctrine Commons expresion builder and implements a visitor for that to build the PHPCR-ODM query.

If this approach seems reasonable I will finish this off and add the tests (PHPSpec or PHPUnit?)

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionBuilder.php:
```diff
@@ -0,0 +1,171 @@
+     */
+    public function like($field, $pattern)
+    {
+        return $this->expressionBuilder->contains($field, $pattern);
```
"like" doesn't exist in the Collections QB API, byt we can actually return a custom comparison as `new Comparison($field, 'like', $pattern)`, same with `notLike`.

dantleech replied:
`like` is analasgous to contains if I understand correctly (it is referenced to as "contains" in the filter).

brunoroux commented on src/Sylius/Bundle/GridBundle/composer.json:
```diff
@@ -25,14 +25,17 @@
         "symfony/form": "^2.7",
-        "symfony/framework-bundle": "^2.7.7"
+        "symfony/framework-bundle": "^2.7.7",
+        "jackalope/jackalope-doctrine-dbal": "^1.2"
```
I think this one should be in require-dev

pamil commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/Driver.php:
```diff
@@ -0,0 +1,52 @@
+ */
+class Driver implements DriverInterface
+{
+    const NAME = 'doctrine/orm';
```
Shouldn't it be `doctrine/phpcr`?

dantleech replied:
yep, `doctrine/phpcr-odm`

pjedrzejewski replied:
👍

pamil commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionBuilder.php:
```diff
@@ -0,0 +1,171 @@
+    private $expressionBuilder;
+    private $orderBys = [];
+
+    function __construct()
```
Missing visibility modifier

pamil commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionBuilder.php:
```diff
@@ -0,0 +1,171 @@
+ */
+class ExpressionBuilder implements ExpressionBuilderInterface
+{
+    private $expressionBuilder;
```
Some typehints would be useful

pamil commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionBuilder.php:
```diff
@@ -0,0 +1,171 @@
+     */
+    public function orderBy($field, $direction)
+    {
+        $this->orderBys = [ [ $field, $direction ] ];
```
Coding standards - spaces around `[` and `]`.

pamil commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionVisitor.php:
```diff
@@ -0,0 +1,137 @@
+        $expressions = $expr->getExpressionList();
+        $node = $baseNode;
+
+        if ($expressionCount = count($expressions) <= 2) {
```
Can we make it more explicit by doing the assignment before the `if`?

pamil commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionVisitor.php:
```diff
@@ -0,0 +1,137 @@
+    /**
+     * {@inheritDoc}
+     */
+    public function walkCompositeExpression(CompositeExpression $expr, AbstractNode $abstractNode)
```
Shouldn't the `$abstractNode` be named `$parentNode`?

pamil commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionVisitor.php:
```diff
@@ -0,0 +1,137 @@
+
+        foreach (array_values($expressions) as $index => $expression) {
+            // if it is the last element dispatch the node itself.
+            if ($index === $expressionCount - 1) {
```
Can we move it outside the loop?

pamil commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionVisitor.php:
```diff
@@ -0,0 +1,137 @@
+        foreach (array_values($expressions) as $index => $expression) {
+            // if it is the last element dispatch the node itself.
+            if ($index === $expressionCount - 1) {
+                $this->dispatch($expression, $node);
```
So for the last methods both `$this->dispatch($expression, $node);` and `$this->dispatch($expression, $node->andX());` should be called?

dantleech replied:
No we should `break` I think - note that this is just very rough code that hasn't been tested.

pamil commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionVisitor.php:
```diff
@@ -0,0 +1,137 @@
+
+    public function dispatch(Expression $expr, $parentNode = null)
+    {
+        if ($parentNode = null) {
```
Assignment?

dantleech replied:
should be `===`

pjedrzejewski commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionBuilder.php:
```diff
@@ -0,0 +1,171 @@
+    public function isNotNull($field)
+    {
+        throw new \BadMethodCallException('Not currently supported.');
+        return $this->expressionBuilder->isNull($field);
```
This line should be removed if we throw exception above.

pjedrzejewski commented:
Hey @dantleech, thanks for this contribution!

If it is really not possible to use PHPCR query builder directly, then this sounds like a smart workaround. I'd say go for it. It does not make sense to add specs for existing code, so spec only new stuff or code that does not work as expected. :)

dantleech commented on src/Sylius/Bundle/GridBundle/Tests/Doctrine/PHPCRODM/ExpressionVisitorTest.php:
```diff
@@ -0,0 +1,224 @@
+use Doctrine\ODM\PHPCR\Query\Builder\OperandStaticLiteral;
+use Doctrine\Common\Collections\Expr\CompositeExpression;
+
+require(__DIR__ . '/QueryBuilderWalker.php');
```
This class creates a simple string representation of the query builder.

dantleech commented:
Updated, we can now build the query for PHPCR-ODM fom the Doctrine Collection Expressions.

I added the test for this with PHPUnit (it is more of a light-weight functional test) and created a simple helper class to generate a string representation of the query builder state (to make testing easier).

Next steps would be to add specs for the other classes and finish this off.

pjedrzejewski commented:
Hey @dantleech, we are going to need this real soon, so I will be happy to help with finishing this. Do you have some time? :)

dantleech commented:
hey @pjedrzejewski , yeah I can work on it tomorrow. I was waiting on https://github.com/Sylius/Sylius/pull/4812

dantleech commented:
Updated, all of the expressions are supported now and manually testing it seems to work fine for the String and Boolean filters. Still some work to do however.

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionBuilder.php:
```diff
@@ -0,0 +1,172 @@
+     */
+    public function comparison($field, $operator, $value)
+    {
+        throw new \BadMethodCallException('Not supported yet.');
```
@pjedrzejewski when does this get used?

pjedrzejewski replied:
Nowhere yet, but I wanted to have it on the interface. Let's keep it for now like for other drivers.

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionBuilder.php:
```diff
@@ -0,0 +1,172 @@
+    public function isNull($field)
+    {
+        // could probably do this by comparing with eq = NULL and handling this case
+        // in the visitor.
```
remove comment

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionVisitor.php:
```diff
@@ -0,0 +1,172 @@
+
+            case Comparison::CONTAINS:
+                return $parentNode->like()->field($this->getField($field))->literal($value)->end();
+            case ExtraComparison::NOT_CONTAINS:
```
Add empty lines as above

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/DataSource.php:
```diff
@@ -0,0 +1,70 @@
+     */
+    public function restrict($expression, $condition = DataSourceInterface::CONDITION_AND)
+    {
+        $visitor = new ExpressionVisitor($this->queryBuilder);
```
Need to implement `$condition` here.

dantleech commented:
Updated - I think this should be OK now. just point aboute the `comparison` method above @pjedrzejewski - although I notice that it is also not implemented for DBAL or ORM.

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/DataSource.php:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * @param QueryBuilder $queryBuilder
+     */
+    function __construct(QueryBuilder $queryBuilder, ExpressionBuilder $expressionBuilder = null)
```
public

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/DataSource.php:
```diff
@@ -0,0 +1,84 @@
+                $parentNode = $this->queryBuilder->orWhere();
+                break;
+            default:
+                throw new \RuntimeException(sprintf(
```
Test this exception

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionBuilder.php:
```diff
@@ -0,0 +1,177 @@
+
+/**
+ * ExpressionBuilder which builds Doctrine Commons
+ * ExpressionBuilder object graph which we can then walk in order
```
Non-sensical.

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionBuilder.php:
```diff
@@ -0,0 +1,177 @@
+     */
+    public function orderBy($field, $direction)
+    {
+        $this->orderBys = [[ $field, $direction ]];
```
Is order by tested?

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionVisitor.php:
```diff
@@ -0,0 +1,175 @@
+use Doctrine\Common\Collections\Expr\Expression;
+use Sylius\Bundle\GridBundle\Doctrine\PHPCRODM\ExtraComparison;
+
+class ExpressionVisitor
```
Doc

dantleech commented on src/Sylius/Bundle/GridBundle/Resources/config/driver/doctrine/phpcr-odm.xml:
```diff
@@ -0,0 +1,32 @@
+    </services>
+
+</container>
+
```
dd

dantleech commented:
Missing "order by" feature + tests and spec/test for ExpressionBuilder

pjedrzejewski commented on src/Sylius/Bundle/GridBundle/Resources/config/driver/doctrine/phpcr-odm.xml:
```diff
@@ -0,0 +1,32 @@
+    </parameters>
+
+    <services>
+        <service id="sylius.grid_driver.doctrine.orm" class="%sylius.grid_driver.doctrine.phpcrodm.class%">
```
Service id is invalid.

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/DataSource.php:
```diff
@@ -0,0 +1,91 @@
+    private $expressionBuilder;
+
+    /**
+     * @param QueryBuilder $queryBuilder
```
MIssing `@param`

dantleech commented on src/Sylius/Bundle/GridBundle/Doctrine/PHPCRODM/ExpressionVisitor.php:
```diff
@@ -0,0 +1,182 @@
+        return $node;
+    }
+
+    public function dispatch(Expression $expr, $parentNode = null)
```
doc

dantleech commented on src/Sylius/Bundle/GridBundle/spec/Doctrine/PHPCRODM/DriverSpec.php:
```diff
@@ -0,0 +1,62 @@
+        ;
+    }
+
+    function it_creates_data_source_via_doctrine_orm_query_builder(
```
s/orm/phpcr-odm

dantleech commented on src/Sylius/Bundle/GridBundle/spec/Doctrine/PHPCRODM/ExpressionBuilderSpec.php:
```diff
@@ -0,0 +1,159 @@
+        $expressionBuilder->neq('o.foo', 'value')->shouldHaveBeenCalled();
+    }
+
+    function it_builds_lessThanOrEqual(
```
unserscores

dantleech commented:
Updated -- implemented order by - so that *should* be the final nail in this PR from a feature POV.

One thing I didn't find out how to test was ordering the grid from the UI - I notice that in the docs there is a `sortable` field in the (`column`) field definition, which seems to be missing from the bundle.

pjedrzejewski commented:
Awesome work, this is a huge help for us, as we will soon add PHPCR stuff to new admin. Thank you Dan! 👍  We will absolutely iterate on this in separate PRs.

Regarding your question, sorting is not yet that functional, I will be looking at it soon for the UI and ORM driver.

I have one question though: why we need to apply sorting in the DataSource and not inside of the ExpressionBuilder?

dantleech commented:
Thanks :)

> why we need to apply sorting in the DataSource

The expression builder is just a factory for `Expr` classes. It returns either `Comparison` or `Composite` (or similar) classes - unlike the Doctrine ORM `ExpressionBuilder` which acts directly upon the ORM `QueryBuilder`. The PHPCR-ODM query builder has no access to the actual query builder.

I just store it in the ExpressionBuilder itself, and we can then later apply it to the actual `PHPCR-ODM` query builder in the `DataSource`.
